### PR TITLE
BUG: df.replace over pd.Period columns (#34871)

### LINKED ISSUE #34871: BUG: Inconsistent behavior for df.replace over pd.Period columns
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd

df = pd.DataFrame({
    'Int': [1, 2, 3],
    'Flo': [1.0, 6.0, 1.0],
    'Per': [pd.Period('2020-01')]*3,
    'Tim': [pd.Timestamp('2020-01')]*3
})

# TypeError: 'value' should be a 'Period', 'NaT', or array of those. Got 'float' instead.
df.replace(1.0, 0.0)

# Works
df[['Int', 'Flo', 'Tim']].replace(1.0, 0.0)

# Works
df['Per'].replace(1.0, pd.Period('2020-02'))

```

#### Problem description

`df.replace(1.0, 0.0)` fails if any columns of `df` are Period types. The error is saying that 0.0 is an invalid value to put in a Pandas Period column. This should never be an issue though, 1.0 will never exist in a Pandas Period column, so nothing will ever be replaced for 0.0.

#### Expected Output

The expected behavior would be consistent with the `df[['Int', 'Flo', 'Tim']].replace(1.0, 0.0)` in the example above. This does work, the 'Flo' column has all of it's 1.0s replaced with 0.0s and the 'Int' and 'Tim' columns (which are not Float types) remain unaffected. I would expect that a Pandas Period column would also be skipped in the same way. 

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.7.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.5.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.4
numpy            : 1.18.5
pytz             : 2020.1
dateutil         : 2.8.0
pip              : 20.1.1
setuptools       : 47.3.0.post20200616
Cython           : None
pytest           : 4.4.1
hypothesis       : None
sphinx           : 3.1.1
blosc            : None
feather          : None
xlsxwriter       : 1.2.9
lxml.etree       : 4.5.1
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.15.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.5.1
matplotlib       : 3.2.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 4.4.1
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
xlsxwriter       : 1.2.9
numba            : None

</details>

### PR DESCRIPTION
**Pull request by samc1213**
- [x] closes #34871
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This commit ensures that PeriodArrays return False for _can_hold_element for any element that is not a pd.Period. This prevents upstream code from casting the dtype to object. Also un-xfail test written in #34904

### CODE REVIEW

**Comment by samc1213** [comment]
This is my first PR, so I would really appreciate guidance on how to improve the commit, and/or feedback on any process I am not following properly. Thanks in advance

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -1975,6 +1976,18 @@ def external_values(self):
         return self.values.astype(object)
 
 
+class PeriodExtensionBlock(ObjectValuesExtensionBlock):
+    """
+    Used by PeriodArray to ensure proper type conversions
+    """
+
+    def _can_hold_element(self, element: Any) -> bool:
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:1984]: this looks good, but is there any way to do this within ObjectValuesExtensionBlock?  (big picture we're looking to get less special-casing of Block subclasses)

so maybe in ObjectValuesExtensionBlock could write

```
def _can_hold_element(self, element):
        tipo = maybe_infer_dtype_type(element)
        if tipo is not None:
            return issubclass(tipo.type, self.dtype.type)
        return isinstance(element, self.dtype.type)
```

?  Might need to check/test that this works for IntervalDtype too (wouldnt surprise me if we see the analogous bug there)

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -10,6 +10,7 @@
 import pandas._libs.internals as libinternals
 from pandas._libs.internals import BlockPlacement
 from pandas._libs.tslibs import conversion
+from pandas._libs.tslibs.period import Period
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:13]: let's just fold this import into the previous line and get it from tslibs

**Comment by jbrockmendel** [comment]
> This is my first PR, so I would really appreciate guidance on how to improve the commit, and/or feedback on any process I am not following properly. Thanks in advance

You're doing well so far.  A couple of comments, then the next thing to do will be to add a note in doc/source/whatsnew/1.2.0.rst giving a short description of the fixed bug

File: pandas/tests/frame/methods/test_replace.py [hunk 3]
```diff
@@ -1517,9 +1517,6 @@ def test_replace_with_duplicate_columns(self, replacement):
 
         tm.assert_frame_equal(result, expected)
 
-    @pytest.mark.xfail(
-        reason="replace() changes dtype from period to object, see GH34871", strict=True
-    )
     def test_replace_period_ignore_float(self):
```
Comment (jbrockmendel) [inline pandas/tests/frame/methods/test_replace.py:1520]: does this also affect/fix Series.replace?  if so let's make sure we have a test for that too

**Comment by samc1213** [comment]
> > This is my first PR, so I would really appreciate guidance on how to improve the commit, and/or feedback on any process I am not following properly. Thanks in advance
> 
> You're doing well so far. A couple of comments, then the next thing to do will be to add a note in doc/source/whatsnew/1.2.0.rst giving a short description of the fixed bug

Thanks for your feedback, addressing your comments now. Also, what is the cause of the build failures? They seem to be happening to everyone, is someone working on fixing them? I was trying to find documentation of the build issue, if it is in fact a widespread thing.

**Comment by jbrockmendel** [comment]
> Also, what is the cause of the build failures? They seem to be happening to everyone, is someone working on fixing them? 

Yes, don't worry about it (already fixed in fact)

**Comment by samc1213** [comment]
Pardon my ignorance, is there anything else I need to address/do to get this PR pulled? Thanks for all your help @jbrockmendel

**Review by jreback — CHANGES_REQUESTED** [review]
can u run some asvs (indexing / reshaping ones)

this could potentially cause some perf issues

File: doc/source/whatsnew/v1.2.0.rst [hunk 4]
```diff
@@ -387,6 +387,11 @@ I/O
 - Removed ``private_key`` and ``verbose`` from :func:`read_gbq` as they are no longer supported in ``pandas-gbq`` (:issue:`34654`, :issue:`30200`)
 - Bumped minimum pytables version to 3.5.1 to avoid a ``ValueError`` in :meth:`read_hdf` (:issue:`24839`)
 
+Period
+^^^^^^
+
+- Bug in :meth:`DataFrame.replace` and :meth:`Series.replace` where :class:`Period` dtypes would be converted to object dytpes (:issue:34871)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:393]: typo on dtypes

File: pandas/tests/frame/methods/test_replace.py [hunk 5]
```diff
@@ -1517,17 +1517,15 @@ def test_replace_with_duplicate_columns(self, replacement):
 
         tm.assert_frame_equal(result, expected)
 
-    @pytest.mark.xfail(
-        reason="replace() changes dtype from period to object, see GH34871", strict=True
-    )
-    def test_replace_period_ignore_float(self):
+    @pytest.mark.parametrize("value", [pd.Period("2020-01"), pd.Interval(0, 5)])
+    def test_replace_period_ignore_float(self, value):
```
Comment (jreback) [inline pandas/tests/frame/methods/test_replace.py:1521]: can you change the text name itself (period -> EA)

File: pandas/tests/series/methods/test_replace.py [hunk 6]
```diff
@@ -449,3 +449,15 @@ def test_replace_with_compiled_regex(self):
         result = s.replace({regex: "z"}, regex=True)
         expected = pd.Series(["z", "b", "c"])
         tm.assert_series_equal(result, expected)
+
+    @pytest.mark.parametrize("value", [pd.Period("2020-01"), pd.Interval(0, 5)])
+    def test_replace_period_ignore_float(self, value):
```
Comment (jreback) [inline pandas/tests/series/methods/test_replace.py:454]: same here

**Review by jreback — CHANGES_REQUESTED** [review]
also there was a reference in the OP to https://github.com/pandas-dev/pandas/issues/35268, if this fixes as well would be great to add the test (if not ok too)

**Comment by samc1213** [comment]
> can u run some asvs (indexing / reshaping ones)
> 
> this could potentially cause some perf issues

Is there a server to run these on? Having issues running these locally in my docker env

**Comment by jbrockmendel** [comment]
> Is there a server to run these on?

No.

> Having issues running these locally in my docker env

are you familiar with using ipython's `%timeit`?

**Comment by jreback** [comment]
can you merge master

**Comment by samc1213** [comment]
> can you merge master

Will do, sorry I've taken so long on this, haven't had much time to try and get the ASVs working

**Comment by samc1213** [comment]
> also there was a reference in the OP to #35268, if this fixes as well would be great to add the test (if not ok too)

It does not fix #35268 - however, I think it's a very similar code change to fix - I can take care of that once this is approved

**Comment by samc1213** [comment]
> > Is there a server to run these on?
> 
> No.
> 
> > Having issues running these locally in my docker env
> 
> are you familiar with using ipython's `%timeit`?

I think I've got this working, will reply with results shortly

**Comment by samc1213** [comment]
> > also there was a reference in the OP to #35268, if this fixes as well would be great to add the test (if not ok too)
> 
> It does not fix #35268 - however, I think it's a very similar code change to fix - I can take care of that once this is approved

Actually, I will modify this branch to fix both, should be a small change

**Comment by samc1213** [comment]
> > > Is there a server to run these on?
> > 
> > 
> > No.
> > > Having issues running these locally in my docker env
> > 
> > 
> > are you familiar with using ipython's `%timeit`?
> 
> I think I've got this working, will reply with results shortly

image

Looks like I'm getting some bad results for indexing ASVs. I'm not sure why this is affecting the NumericSeriesIndexing tests, this shouldn't be affecting numeric series?

**Comment by jbrockmendel** [comment]
> Looks like I'm getting some bad results for indexing ASVs. I'm not sure why this is affecting the NumericSeriesIndexing tests, this shouldn't be affecting numeric series?

Unfortunately its not unusual for the asvs to include some noise.  Are these results consistent across runs?  (just re-run the surprising outliers)

**Comment by samc1213** [comment]
> > Looks like I'm getting some bad results for indexing ASVs. I'm not sure why this is affecting the NumericSeriesIndexing tests, this shouldn't be affecting numeric series?
> 
> Unfortunately its not unusual for the asvs to include some noise. Are these results consistent across runs? (just re-run the surprising outliers)

Thanks, looks like I'm making it faster now 😄 

image

File: pandas/core/internals/blocks.py [hunk 7]
```diff
@@ -2002,6 +2002,18 @@ class ObjectValuesExtensionBlock(ExtensionBlock):
     def external_values(self):
         return self.values.astype(object)
 
+    def _can_hold_element(self, element: Any) -> bool:
+        if is_valid_nat_for_dtype(element, self.dtype):
+            return True
+        if element is NaT:
```
Comment (jreback) [inline pandas/core/internals/blocks.py:2008]: @jbrockmendel ok here?

@samc1213 are all of these branches actually hit? in particular why L2008 and L2010

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py (hunk 7 above)
Comment (samc1213) [reply pandas/core/internals/blocks.py]: @jreback Added these due to unit tests. Specifically, L2010 is required for `pandas/tests/extension/test_interval.py::TestSetitem::test_setitem_empty_indxer` to pass. L2008 may not be required, I just removed it and will see if the tests pass...

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @jbrockmendel if you'd look and merge if ok.

**Comment by jreback** [comment]
cc @jbrockmendel ok here

**Comment by jbrockmendel** [comment]
will take a look this afternoon

File: doc/source/whatsnew/v1.2.0.rst [hunk 8]
```diff
@@ -434,6 +434,8 @@ Strings
 
 Interval
 ^^^^^^^^
+
+- Bug in :meth:`DataFrame.replace` and :meth:`Series.replace` where :class:`Interval` dtypes would be converted to object dtypes (:issue:34871)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.2.0.rst:438]: backticks around the issue number

Comment (jbrockmendel) [reply doc/source/whatsnew/v1.2.0.rst:438]: same below

File: doc/source/whatsnew/v1.2.0.rst (hunk 4 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v1.2.0.rst:496]: > dytpes

typo is still there.  BTW in general let the person who made the original comment hit the "conversation resolved" button

File: pandas/core/internals/blocks.py [hunk 9]
```diff
@@ -2002,6 +2002,16 @@ class ObjectValuesExtensionBlock(ExtensionBlock):
     def external_values(self):
         return self.values.astype(object)
 
+    def _can_hold_element(self, element: Any) -> bool:
+        if is_valid_nat_for_dtype(element, self.dtype):
+            return True
+        if isinstance(element, list) and len(element) == 0:
+            return True
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:2009]: why this? i dont think we have this in other _can_hold_element methods

File: pandas/tests/series/methods/test_replace.py [hunk 10]
```diff
@@ -450,6 +450,18 @@ def test_replace_with_compiled_regex(self):
         expected = pd.Series(["z", "b", "c"])
         tm.assert_series_equal(result, expected)
 
+    @pytest.mark.parametrize("value", [pd.Period("2020-01"), pd.Interval(0, 5)])
+    def test_replace_ea_ignore_float(self, value):
+        """
+        Regression test for corrolary to GH#34871: if series.replace(1.0, 0.0)
+        is called on a Period/Interval Series, the old, faulty behavior
+        is to raise TypeError.
+        """
```
Comment (jbrockmendel) [inline pandas/tests/series/methods/test_replace.py:459]: this can be a comment, not a docstring, just needs to see `# GH#34871`, ditto above

Comment (jbrockmendel) [reply pandas/tests/series/methods/test_replace.py:459]: actually you can share this test using the frame_or_series fixture

File: pandas/core/internals/blocks.py (hunk 9 above)
Comment (samc1213) [reply pandas/core/internals/blocks.py:2009]: Without it, `pandas/tests/extension/test_interval.py::TestSetitem::test_setitem_empty_indexer` will not pass. It is also used on L2223. If there's a better way to handle this please let me know, but I did borrow this from 2223

File: doc/source/whatsnew/v1.2.0.rst (hunk 8 above)
Comment (samc1213) [reply doc/source/whatsnew/v1.2.0.rst:438]: Thanks, fixed

File: doc/source/whatsnew/v1.2.0.rst (hunk 4 above)
Comment (samc1213) [reply doc/source/whatsnew/v1.2.0.rst:496]: Gotcha thanks for the heads up, haven't used github at all really. Fixed

File: pandas/tests/series/methods/test_replace.py (hunk 10 above)
Comment (samc1213) [reply pandas/tests/series/methods/test_replace.py:459]: Changed to a comment, but I grepped for frame_or_series and not finding much in the codebase. I did find `index_or_series`. Can you please elaborate more on how to use frame_or_series? There do seem to be a fair amount of duplication between these two files (series/test_replace and frame/test_replace). `test_replace_with_compiled_regex` is basically copy-pasted between the two, for example

Comment (arw2019) [reply pandas/tests/series/methods/test_replace.py]: > Can you please elaborate more on how to use frame_or_series? 

It's a pytest fixture:
https://github.com/pandas-dev/pandas/blob/aa390ecc825464953f91e9214ed65ef5f83be5d6/pandas/conftest.py#L296

Comment (samc1213) [reply pandas/tests/series/methods/test_replace.py]: Thanks @arw2019  - I needed to merge master in locally to find this

File: pandas/tests/generic/methods/test_replace.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+        expected = frame_or_series([value] * 3)
+        tm.assert_equal(expected, result)
+
+    def test_replace_with_compiled_regex(self):
+        # https://github.com/pandas-dev/pandas/issues/35680
+        s = pd.Series(["a", "b", "c"])
+        regex = re.compile("^a$")
+        result = s.replace({regex: "z"}, regex=True)
+        expected = pd.Series(["z", "b", "c"])
+        tm.assert_series_equal(result, expected)
```
Comment (jbrockmendel) [inline pandas/tests/generic/methods/test_replace.py:24]: tests only belong in tests.generic.methods if _all_ tests for that method are fully parametrized over DataFrame/Series, which these are not.

(i think this is confusing, so will likely just get rid of this directory at some point

File: pandas/tests/frame/methods/test_replace.py [hunk 12]
```diff
@@ -1610,14 +1597,6 @@ def test_replace_dict_category_type(self, input_category_df, expected_category_d
 
         tm.assert_frame_equal(result, expected)
 
-    def test_replace_with_compiled_regex(self):
```
Comment (jbrockmendel) [inline pandas/tests/frame/methods/test_replace.py:1613]: this is unrelated to the current PR, right?  if you want to make a separate PR to parametrize this test, go for it.  better to just use frame_or_series in this file and remove the version in the series tests

File: pandas/tests/generic/methods/test_replace.py (hunk 11 above)
Comment (samc1213) [reply pandas/tests/generic/methods/test_replace.py:24]: I see, makes sense then. I'll just test frame and series in the same file then

File: pandas/tests/frame/methods/test_replace.py (hunk 12 above)
Comment (samc1213) [reply pandas/tests/frame/methods/test_replace.py:1613]: Sorry, yeah its unrelated. Undone, and just used frame_or_series here, just trying to organize it properly (evidently was misguided :D )

Comment (jbrockmendel) [reply pandas/tests/frame/methods/test_replace.py]: > evidently was misguided

not at all, organizing/parametrizing/breaking-large-ancient-tests-into-specific-tests is something we very much want to encourage.  just in dedicated PRs

File: pandas/tests/frame/methods/test_replace.py [hunk 13]
```diff
@@ -1517,18 +1517,13 @@ def test_replace_with_duplicate_columns(self, replacement):
...
-        """
-        Regression test for GH#34871: if df.replace(1.0, 0.0) is called on a df
-        with a Period column the old, faulty behavior is to raise TypeError.
-        """
-        df = DataFrame({"Per": [pd.Period("2020-01")] * 3})
+    @pytest.mark.parametrize("value", [pd.Period("2020-01"), pd.Interval(0, 5)])
+    def test_replace_ea_ignore_float(self, frame_or_series, value):
+        # GH#34871
+        df = frame_or_series([value] * 3)
         result = df.replace(1.0, 0.0)
```
Comment (jbrockmendel) [inline pandas/tests/frame/methods/test_replace.py:1524]: since this is now frame_or_series, can you call it `obj` instead of `df`

**Comment by jbrockmendel** [comment]
One last nitpick, otherwise LGTM, cc @jreback

File: pandas/tests/frame/methods/test_replace.py (hunk 13 above)
Comment (samc1213) [reply pandas/tests/frame/methods/test_replace.py:1524]: Got it, done

File: pandas/tests/frame/methods/test_replace.py (hunk 12 above)
Comment (samc1213) [reply pandas/tests/frame/methods/test_replace.py]: I see, definitely makes sense - thanks

**Comment by jreback** [comment]
thanks @samc1213

**Comment by samc1213** [comment]
@jreback  @jbrockmendel Thanks so much for your patience and feedback, this was a cool experience for me. Hopefully I'll be back! Appreciate all the time you put into this great project
